src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/DefaultRazorProjectService.cs

```diff
@@ -188,8 +188,21 @@ public override void RemoveDocument(string filePath)
         }
 
         var defaultProject = (ProjectSnapshot)projectSnapshot;
-        _logger.LogInformation("Removing document '{textDocumentPath}' from project '{projectSnapshotFilePath}'.", textDocumentPath, projectSnapshot.FilePath);
-        _projectSnapshotManagerAccessor.Instance.DocumentRemoved(defaultProject.HostProject, documentSnapshot.State.HostDocument);
+
+        // If the document is open, we can't remove it, because we could still get a request for it, and that
+        // request would fail. Instead we move it to the miscellaneous project, just like if we got notified of
+        // a remove via the project.razor.json
+        if (_projectSnapshotManagerAccessor.Instance.IsDocumentOpen(textDocumentPath))
+        {
+            _logger.LogInformation("Moving document '{textDocumentPath}' from project '{projectSnapshotFilePath}' to misc files because it is open.", textDocumentPath, projectSnapshot.FilePath);
+            var miscellaneousProject = (ProjectSnapshot)_projectResolver.GetMiscellaneousProject();
+            MoveDocument(textDocumentPath, defaultProject, miscellaneousProject);
```

Comment:
At some point do we move the document from the misc workspace to the main workspace if that same document is recreated?

Reply:
Yes. When we get project info from the project.razor.json we move things to the misc project exactly like this, and we restore them to the real project here: https://github.com/davidwengier/razor/blob/FixDeleteFileIssue/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/DefaultRazorProjectService.cs#L383